package furious;

option optimize_for = LITE_RUNTIME;

enum DataType {
	FLOAT64 = 0;
	FLOAT32 = 1;
}

enum TriangularMatrixType {
	UPPER = 0;
	LOWER = 1;
}

enum TranspositionType {
	NORMAL    = 0;
	TRANSPOSE = 1;
}

message Request {
	enum Type {
		EMPTY_ARRAY              =  0;
		DATA_ARRAY               =  1;
		CONST_ARRAY              =  2;
		IDENTITY_MATRIX          =  3;
		LINSPACE                 =  4;
		RESHAPE                  =  5;
		REPEAT                   =  6;
		DEALLOCATE               =  7;
		FETCH                    =  8;
		BARRIER                  =  9;
		INFO                     = 10;
		BINARY_OPERATION         = 11;
		BINARY_CONST_OPERATION   = 12;
		UNARY_OPERATION          = 13;
		REDUCTION_OPERATION      = 14;
		AXIS_REDUCTION_OPERATION = 15;
		DOT_OPERATION            = 16;
		CHOLESKY_DECOMPOSITION   = 17;
		SOLVE_TRIANGULAR         = 18;
	}
	required fixed32                      id                             =  1;
	required Type                         type                           =  2;

	optional EmptyArrayRequest            empty_array_request            =  3;
	optional DataArrayRequest             data_array_request             =  4;
	optional ConstArrayRequest            const_array_request            =  5;
	optional IdentityMatrixRequest        identity_matrix_request        =  6;
	optional LinspaceRequest              linspace_request               =  7;
	optional ReshapeRequest               reshape_request                =  8;
	optional RepeatRequest                repeat_request                 =  9;
	optional DeallocateRequest            deallocate_request             = 10;
	optional FetchRequest                 fetch_request                  = 11;
	optional BinaryOperationRequest       binary_operation_request       = 12;
	optional BinaryConstOperationRequest  binary_const_operation_request = 13;
	optional UnaryOperationRequest        unary_operation_request        = 14;
	optional ReductionRequest             reduction_request              = 15;
	optional AxisReductionRequest         axis_reduction_request         = 16;
	optional DotOperationRequest          dot_operation_request          = 17;
	optional CholeskyDecompositionRequest cholesky_decomposition_request = 18;
	optional SolveTriangularRequest       solve_triangular_request       = 19;
}

message EmptyArrayRequest {
	required fixed32  id_out      = 1;
	repeated uint32   shape       = 2 [packed=true];
	required DataType data_type   = 3;
}

message DataArrayRequest {
	required fixed32  id_out      = 1;
	repeated uint32   shape       = 2 [packed=true];
	required DataType data_type   = 3;
	required bytes    data_buffer = 4;
}

message ConstArrayRequest {
	required fixed32  id_out      = 1;
	repeated uint32   shape       = 2 [packed=true];
	required DataType data_type   = 3;
	required double   fill_value  = 4;
}

message IdentityMatrixRequest {
	required fixed32  id_out      = 1;
	required fixed32  rows        = 2;
	required fixed32  columns     = 3;
	required sfixed32 diagonal    = 4;
	required DataType data_type   = 5;
}

message LinspaceRequest {
	required sfixed32  id_out     = 1;
	required double    start      = 2;
	required double    stop       = 3;
	required uint32    samples    = 4;
	required bool      closed     = 5;
	required DataType  data_type  = 6;
}

message ReshapeRequest {
	required sfixed32  id_a      = 1;
	required fixed32   id_out    = 2;
	repeated uint32    shape_out = 3 [packed=true];
}

message RepeatRequest {
	required sfixed32 id_a    = 1;
	required fixed32  id_out  = 2;
	required uint32   axis    = 3;
	required uint32   repeats = 4;
}

message DeallocateRequest {
	required fixed32 id_a = 1;
}

message FetchRequest {
	required sfixed32 id_a = 1;
}

message BinaryOperationRequest {
	enum Type {
		ADD = 0;
		SUB = 1;
		MUL = 2;
		DIV = 3;
	}
	required Type     type   = 1;
	required sfixed32 id_a   = 2;
	required sfixed32 id_b   = 3;
	required fixed32  id_out = 4;
}

message BinaryConstOperationRequest {
	enum Type {
		ADDC  = 0;
		SUBC  = 1;
		SUBRC = 2;
		MULC  = 3;
		DIVC  = 4;
		DIVRC = 5;
	}
	required Type     type    = 1;
	required sfixed32 id_a    = 2;
	required double   value_b = 3;
	required fixed32  id_out  = 4;
}

message UnaryOperationRequest {
	enum Type {
		NEG    = 0;
		ABS    = 1;
		EXP    = 2;
		LOG    = 3;
		SQRT   = 4;
		SQUARE = 5;
	}
	required Type     type   = 1;
	required sfixed32 id_a   = 2;
	required fixed32  id_out = 3;
}

message ReductionRequest {
	enum Type {
		SUM = 0;
		MIN = 1;
		MAX = 2;
	}
	required Type     type   = 1;
	required sfixed32 id_a   = 2;
	required fixed32  id_out = 3;
}

message AxisReductionRequest {
	enum Type {
		SUM = 0;
		MIN = 1;
		MAX = 2;
	}
	required Type     type   = 1;
	required sfixed32 id_a   = 2;
	required uint32   axis   = 3;
	required fixed32  id_out = 4;
}

message DotOperationRequest {
	required sfixed32 id_a   = 1;
	required sfixed32 id_b   = 2;
	required fixed32  id_out = 3;
}

message CholeskyDecompositionRequest {
	required sfixed32             id_a   = 1;
	required TriangularMatrixType a_type = 2;
	required fixed32              id_out = 3;
}

message SolveTriangularRequest {
	required sfixed32             id_a            = 1;
	required TriangularMatrixType a_type          = 2;
	required TranspositionType    a_transposition = 3;
	required bool                 unit_diagonal   = 4;
	required sfixed32             id_y            = 5;
	required fixed32              id_x            = 6;
}
